### PR TITLE
WIP: Add bindings for siphash-2-4

### PR DESCRIPTION
**Pull request by lmctv**
also known as shorthash in libsodium circles...

### CODE REVIEW

**Comment by reaperhulk** [comment]
@lmctv you're fine asking for help/reviews in #cryptography-dev btw. cc'ing @warner to see what he thinks about this :)

**Comment by warner** [comment]
Seems reasonable to me. I'm a bit confused about the difference between "shorthash" and "siphash", but it looks like libsodium is using `src/libsodium/crypto_shorthash/siphash24/*`, so following whatever they do is appropriate.

(which reminds me, we should circle back around to renaming pynacl to pysodium. I'm ok with making bindings to anything that's already in libsodium).

**Comment by lmctv** [comment]
@warner sodium does "functional" renaming of the primitives, but still leaves thei original names available; I prefer keeping both the original and the renamed entry points for clarity.

File: src/nacl/bindings/crypto_shorthash.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    :param data:
+    :type data: bytes
+    :param key:
+    :type key: bytes
+    """
+    _key = ffi.new("unsigned char[]", KEYBYTES)
+    digest = ffi.new("unsigned char[]", BYTES)
+    ffi.memmove(_key, key, min(len(key), KEYBYTES))
+    # The key will be truncated or zero-padded to KEYBYTES length
```
Comment (reaperhulk) [inline src/nacl/bindings/crypto_shorthash.py]: Is this a common convention? It seems odd to me to not just require that `len(key) == KEYBYTES`

File: src/nacl/bindings/crypto_shorthash.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    rc = lib.crypto_shorthash_siphash24(digest, data, len(data), _key)
+
+    assert rc == 0
```
Comment (reaperhulk) [inline src/nacl/bindings/crypto_shorthash.py]: This is fine for now, but #212 applies for the future.

File: tests/test_shorthash.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+
+import pytest
+
+from nacl.bindings import crypto_shorthash_siphash24
+
+HASHES = [
```
Comment (reaperhulk) [inline tests/test_shorthash.py]: Where do these come from? Can you add a comment explaining their provenance?

Comment (lmctv) [inline tests/test_shorthash.py]: Added, it's in 6459e3b .

File: src/nacl/bindings/crypto_shorthash.py (hunk 1 above)
Comment (lmctv) [inline src/nacl/bindings/crypto_shorthash.py]: I've just been following what I'd done for blake2b, whose key parameter is zero padded.

**Comment by lmctv** [comment]
@reaperhulk  thank you very much for your first review. Let me know if you prefer this not-squashed commit sequence or a squashing rebase instead.

File: src/nacl/bindings/crypto_shorthash.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+def crypto_shorthash_siphash24(data, key):
+    """Compute a fast, cryptographic quality, keyed hash of the input data
+
+    :param data:
+    :type data: bytes
+    :param key: len(key) must be equal to
+                :py:data:`.KEYBYTES` (16)
+    :type key: bytes
+    """
+    assert len(key) == KEYBYTES
```
Comment (reaperhulk) [inline src/nacl/bindings/crypto_shorthash.py]: Let's make this `if len(key) != KEYBTES: raise ValueError("Key length must be {0} bytes".format(KEYBYTES))` or something like that. Then we can update the test to look for ValueError

Comment (lmctv) [inline src/nacl/bindings/crypto_shorthash.py]: Changed as requested. Thank you once more.

**Pull request merged**